### PR TITLE
IDEX-2971: Tree collapse fixes

### PR DESCRIPTION
**Pull request by vzhukovskii**
@vparfonov

### CODE REVIEW

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/project/node/HasStorablePath.java [hunk 1]
```diff
@@ -16,6 +16,22 @@
  * @author Vlad Zhukovskiy
  */
 public interface HasStorablePath {
+
```
Comment (skabashnyuk) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/project/node/HasStorablePath.java]: javadoc for class

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/explorer/project/NewProjectExplorerView.java [hunk 2]
```diff
@@ -51,11 +48,7 @@
 
     void setFoldersAlwaysOnTop(boolean foldersAlwaysOnTop);
 
-    void synchronizeTree();
-
-    HandlerRegistration addBeforeExpandNodeHandler(BeforeExpandNodeHandler handler);
-
-    void reloadChildren(List<Node> nodes, Object selectAfter, boolean callAction, boolean goInto);
+    void reloadChildren(Node parent, HasDataObject<?> selectAfter, boolean actionPerformed, boolean goInto);
```
Comment (skabashnyuk) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/explorer/project/NewProjectExplorerView.java]: javadoc for entire class

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/explorer/project/NewProjectExplorerViewImpl.java [hunk 3]
```diff
@@ -176,38 +185,30 @@ public void onClick(ClickEvent event) {
                 }
             }
         }, ClickEvent.getType());
+
+        goIntoHandler = new GoIntoHandler();
+        tree.addExpandHandler(goIntoHandler);
```
Comment (skabashnyuk) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/explorer/project/NewProjectExplorerViewImpl.java]: may be goIntoHandler can be inlined

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/explorer/project/NewProjectExplorerViewImpl.java [hunk 4]
```diff
@@ -388,71 +337,44 @@ public void onClick(ClickEvent event) {
...
     }
 
-    @Override
-    public void synchronizeTree() {
-        tree.synchronize();
+    private void hideProjectInfo() {
+        goIntoBackButton = null;
+        menuPanel.clear();
+        toolBar.remove(projectHeader);
+        setToolbarHeight(22);
```
Comment (skabashnyuk) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/explorer/project/NewProjectExplorerViewImpl.java]: 22 is some magic value? Are you sure it's 22?

File: ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/TreeNodeLoader.java [hunk 5]
```diff
@@ -100,6 +165,60 @@ public void onBeforeLoad(BeforeLoadEvent event) {
         }
     }
 
+    private boolean isNodeHasSameDataObject(Node node, HasDataObject<?> dataNode) {
```
Comment (skabashnyuk) [inline ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/TreeNodeLoader.java]: this method looks like a shit. Sorry.

**Comment by vparfonov** [comment]
Agree with Sergey need add javadoc and i don't see any test. I will merge it in case we need make release ASAP. 
@vzhukovskii  fix comment ASAP

File: ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/TreeNodeLoader.java (hunk 5 above)
Comment (skabashnyuk) [inline ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/TreeNodeLoader.java]: all three in line

**Comment by vparfonov** [comment]
Also we have filed test on travis

File: ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/event/LoadEvent.java [hunk 6]
```diff
@@ -44,12 +45,19 @@
...
     private List<Node> receivedNodes;
+    private LoadParams params;
 
     public LoadEvent(Node requestedNode, List<Node> receivedNodes) {
         this.requestedNode = requestedNode;
         this.receivedNodes = receivedNodes;
     }
 
+    public LoadEvent(Node requestedNode, List<Node> receivedNodes, LoadParams params) {
+        this.requestedNode = requestedNode;
```
Comment (skabashnyuk) [inline ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/event/LoadEvent.java]: Why one constructor is not calling bigger one?

File: ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/TreeNodeLoader.java [hunk 7]
```diff
@@ -57,7 +65,7 @@
      *
      * @see org.eclipse.che.ide.api.project.node.interceptor.NodeInterceptor
      */
-    private Set<NodeInterceptor> nodeInterceptors;
+    private List<NodeInterceptor> nodeInterceptors;
```
Comment (skabashnyuk) [inline ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/TreeNodeLoader.java]: Why don't we use TreeSet here?

File: ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/TreeNodeLoader.java [hunk 8]
```diff
@@ -86,7 +94,64 @@ public void onLoad(LoadEvent event) {
                 tree.getView().onJointChange(tree.findNode(parent), Tree.Joint.NONE);
             }
 
-            tree.getNodeStorage().replaceChildren(parent, event.getReceivedNodes());
+            NodeDescriptor requested = tree.findNode(parent);
+
+            if (requested == null) {
+                //smth happened, that requested node isn't registered in storage
+                Log.info(this.getClass(), "Requested node not found.");
```
Comment (skabashnyuk) [inline ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/TreeNodeLoader.java]: Are you sure you need info here?

File: ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/TreeNodeLoader.java [hunk 9]
```diff
@@ -86,7 +94,64 @@ public void onLoad(LoadEvent event) {
...
+
+            //now search new nodes to add then into the tree
+            List<Node> newNodes = findNewNodes(requested, event.getReceivedNodes());
+
+            if (removedNodes.isEmpty() && newNodes.equals(event.getReceivedNodes())) {
+                tree.getNodeStorage().replaceChildren(parent, newNodes);
+            } else {
+                for (NodeDescriptor removed : removedNodes) {
+                    if (!tree.getNodeStorage().remove(removed.getNode())) {
+                        Log.info(this.getClass(), "Failed to remove node: " + removed.getNode().getName());
```
Comment (skabashnyuk) [inline ide/che-core-ide-ui/src/main/java/org/eclipse/che/ide/ui/smartTree/TreeNodeLoader.java]: Are you sure you need info here?

**Pull request merged**